[JENKINS-61687] Prevent concurrent build deletion

gmshake opened the pull request:
See [JENKINS-61687](https://issues.jenkins-ci.org/browse/JENKINS-61687).

### Proposed changelog entries

* ENKINS-61687: Prevent concurrent build deletion

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@daniel-beck
@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

gmshake committed:
[JENKINS-61687] Prevent concurrent build deletion

stellargo commented on core/src/main/java/hudson/model/Run.java:
```diff
@@ -1612,6 +1616,14 @@ public synchronized void deleteArtifacts() throws IOException {
     public void delete() throws IOException {
+        synchronized (this) {
+            // Avoid concurrent delete.
+            if (state == State.PENDING_DELETE)
+                return;
```
```suggestion
            if (state == State.PENDING_DELETE) {
                return;
            }
```

gmshake replied:
Fixed :)

stellargo commented on core/src/main/java/hudson/model/Run.java:
```diff
@@ -1612,6 +1616,14 @@ public synchronized void deleteArtifacts() throws IOException {
      */
     public void delete() throws IOException {
+        synchronized (this) {
+            // Avoid concurrent delete.
```
```suggestion
            // Avoid concurrent delete. See https://issues.jenkins-ci.org/browse/JENKINS-61687
```

gmshake replied:
Fixed :)

stellargo reviewed (approved):
Only slight nit pick. Implementation seems thread safe to me and looks like it should fix the bug referenced in Jira. But I have not tested it locally.

gmshake commented:
Changed as @stellargo suggested, and rebased to lastest upstream master.

daniel-beck commented:
Thanks for your submission! This is an interesting approach.

I wonder whether it's possible to delete builds in progress, if so, this may mess things up for other uses of the `state`.

Still, this doesn't really add new API (worst case a new, immediately `@Deprecated` status), so it might be worth a try.

gmshake commented:
A second approach would be that, we do not pollute `state` but introduce a boolean flag called `isPendingDelete`. The `pendingDelete` guards only the `delete()` method.

I'm not sure which one is better, but looks current logic in `delete()` does not consider builds those in progress.
So it might be problematical if a long running build would be deleted before it complete.

daniel-beck commented:
Good point, that's probably safer to do and won't interfere with any public code handling `Run.State`, can be `private` or `protected`.

I wonder whether all major job implementations (Pipeline; Matrix; Maven, both module and parent; Freestyle) go through `Run#delete` though (i.e. at least calling the `super`). Something to check.

gmshake commented:
Updated and use flag `isPendingDelete` . Local test looks good. :)

oleg-nenashev reviewed (commented):
LGTM. In some cases it may to run deletion not happening due to error while API callers will get a "success" result. Maybe it makes sense to actually lock requests and propagate errors to all callers if an issue happens. Without async API and timeouts such implementation would be error-prone on its own, so maybe the fix is acceptable as a simple patch

gmshake commented:
I aggree it's just a simple fix only preventing concurrent deleting.

The current state:
 ![current state](http://www.plantuml.com/plantuml/svg/NT3BIiOm40NWVPvYra9nxq9gBqYXDgLDLo8b9bC9JLFD1VJjJHJwbswooNTcJf8xpwqFcoPkLUQJMxw1ypyQuscL-BiXhtTla6MFq54sZ-nfO7K53o0iSeyIxk4hO41vi9A-z6tzZ_JE9Instq0gjvBpX4IBgMchfdkE4hzH18znt6x6LAe8IajbFiY1KgAd8vlxWLPJoHhQnU1xSk224EXSZ2zSwM2HNEMrsdbXRUInxcPI8ZwAv6Xa_8XV)

After the fix:
 ![after fix](http://www.plantuml.com/plantuml/svg/jP0_QyCm5CJtzodUN0ZP3Ibf8b42gMGIPoebkDPB4P7bLd-WfUbthvI0SKo6Bbqyd7-dktTp7seNOcl07xLzhrtTeWzVXjBNwURy1vxlNd0okKSf5hlLcgt48oHRe860tpqbPBNVLW-RYZCii76KI8LJ_8WKIVNOKZwLQtw1J5ShFBYkHQNz4OOlPBGFAP0-gOafUOgTtURgYwYDqlODUYWxIhcjzkL6ijsoMacHZA-PGn-RXhnFzaEjJNG4Ot9SRHZ8hUhZu7I2gvDBBXXF-Nl6inTkQ6Tsh1Meo50W-3SOLMS9RXOgKFkIRBw8NTZPB4rZFC6VqE0YmGz8SRL-cWNc2Osj-GK0)

gmshake commented:
It looks more reasonable like this one, but requires introducing two new states : 
![aaa](http://www.plantuml.com/plantuml/svg/bOzDIyCm6CVl-HJFMHZUFSZS4gGmcx9c9v4IypoJi3JLl81-UnC7jVIsImZ__FulsSQaGygZO_5Y_QSEUeIOVXoLCrZpzy9UxzvWitc4LggXLqz79JWyW0ca4o7SmrUcJ3XZU_dI7SGLSfD6E8Tf1BJnmfO95TsTcWDlskT2qZUPdAZ4JRwletRPEhJ-WyrGTNIoLqDtbFoqLuriY_6zSX2pCHHZiP-rTJaGMvFhQSj2uN6k-sUgetngfNIYhMi6BgfMWf0S9M8h_KRqUkKp0pXvObLTBjcInpowNm00)

The `Run#delete()` need some state check like
```java
    /**
     * Deletes this build and its entire log
     *
     * @throws IOException
     *      if we fail to delete.
     */
    public void delete() throws IOException {
        synchronized (this) {
            if (state == State.DELETED) {
                return; // FIXME noop, or throw exception to indicate this build has been deleted ??
            }
            if (state == State.PENDING_DELETE) {
                return; // FIXME or throw exception to indicate concurrent deletion ??
            }
            if (state != State.COMPLETED) {
                // It looks not good to delete ongoing build.
                // FIXME or we cancel this build, then delete again ?
                throw new IllegalStateException("cannot delete build while in " + state);
            }

            state = State.PENDING_DELETE;
        }
        // It's safe to delete now.
        // Deleting process omitted.
        ...
       

        // Set state after deleting
        synchronized (this) {
            state = State.DELETED
        }
    }
```

oleg-nenashev commented:
Thanks for the analysis @gmshake ! I think that the 2-state approach is a good solution here. It would have been an overkill to add states to the graph if `State` was a public API, but fortunately the `State` is a private class. I am +1 for the suggested solution, though it might require significant changes within the class to ensure the new states are properly interpreted everywhere. If you do not want to work on it right now, I think we could merge the current PR and create a follow-up task. From what I can tell, this PR is a net improvement anyway

gmshake commented:
@oleg-nenashev  Concurrency is hard. More work is required for the suggested solution.

> I think we could merge the current PR and create a follow-up task. 

I'm OK to merge the current PR :)

oleg-nenashev reviewed (approved):
I have created https://issues.jenkins-ci.org/browse/JENKINS-63275 as a follow-up.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process
